backend/src/stwo/circuit_builder.rs

```diff
@@ -0,0 +1,204 @@
+                (name.clone(), values)
+            })
+            .collect(),
+    );
+
+    let domain = CanonicCoset::new(analyzed.degree().ilog2()).circle_domain();
+    element
+        .map(|elements| {
+            elements
+                .iter()
+                .map(|(_, base_column)| CircleEvaluation::new(domain, base_column.clone()))
+                .collect()
+        })
+        .unwrap()
+}
+
+pub struct PowdrEval<T> {
+    analyzed: Arc<Analyzed<T>>,
+    witness_columns: BTreeMap<PolyID, usize>,
+}
+
+impl<T: FieldElement> PowdrEval<T> {
+    pub fn new(analyzed: Arc<Analyzed<T>>) -> Self {
+        let witness_columns: BTreeMap<PolyID, usize> = analyzed
+            .definitions_in_source_order(PolynomialType::Committed)
+            .flat_map(|(symbol, _)| symbol.array_elements())
+            .enumerate()
+            .map(|(index, (_, id))| (id, index))
+            .collect();
+
+        Self {
+            analyzed,
+            witness_columns,
+        }
+    }
+}
+
+impl<T: FieldElement> FrameworkEval for PowdrEval<T> {
+    fn log_size(&self) -> u32 {
+        self.analyzed.degree().ilog2()
+    }
+    fn max_constraint_log_degree_bound(&self) -> u32 {
+        self.analyzed.degree().ilog2() + 1
+    }
+    fn evaluate<E: EvalAtRow>(&self, mut eval: E) -> E {
+        assert!(
+            self.analyzed.constant_count() == 0 && self.analyzed.publics_count() == 0,
+            "Error: Expected no fixed columns nor public inputs, as they are not supported yet.",
+        );
+
+        let col_count = self.analyzed.commitment_count();
+        let mut witness_eval = Vec::with_capacity(col_count);
+        for _ in 0..col_count {
+            witness_eval.push(eval.next_interaction_mask(0, [0, 1]));
+        }
+
+        for id in self
+            .analyzed
+            .identities_with_inlined_intermediate_polynomials()
+        {
+            match id {
+                Identity::Polynomial(identity) => {
+                    let expr = to_stwo_expression::<T, E>(
+                        &self.witness_columns,
+                        &identity.expression,
+                        &witness_eval,
+                    );
+                    eval.add_constraint(expr);
+                }
+                Identity::Connect(..) => {
+                    unimplemented!("Connect is not implemented in stwo yet")
+                }
+                Identity::Lookup(..) => {
+                    unimplemented!("Lookup is not implemented in stwo yet")
+                }
+                Identity::Permutation(..) => {
+                    unimplemented!("Permutation is not implemented in stwo yet")
+                }
+                Identity::PhantomPermutation(..) => {
+                    unimplemented!("PhantomPermutation is not implemented in stwo yet")
+                }
+                Identity::PhantomLookup(..) => {
+                    unimplemented!("PhantomLookup is not implemented in stwo yet")
+                }
```

Comment:
These two should not panic, but rather do nothing, since these identities are only used in witness generation

Reply:
Done